dashboard/src/scenes/organisation/DragAndDropSettings.js

```diff
@@ -0,0 +1,260 @@
+import React, { useState, useEffect, useRef, useCallback, useMemo } from 'react';
+import SortableJS from 'sortablejs';
+import { toast } from 'react-toastify';
+import ButtonCustom from '../../components/ButtonCustom';
+import { ModalContainer, ModalBody, ModalFooter, ModalHeader } from '../../components/tailwind/Modal';
+import { capture } from '../../services/sentry';
+
+/*
+data: [
+  {
+    groupTitle: string,
+    items: [any]
+  }
+]
+*/
+
+const DragAndDropSettings = ({
+  title,
+  data,
+  onDragAndDrop,
+  addButtonCaption,
+  ItemComponent,
+  onGroupTitleChange,
+  NewItemComponent,
+  dataItemKey = (item) => item._id,
+  sectionId = 'drag-and-drop-setting',
+  onAddGroup = null,
+  onDeleteGroup = null,
+}) => {
+  if (!title) throw new Error('title is required');
+  if (!data) throw new Error('data is required');
+  if (!onDragAndDrop) throw new Error('onDragAndDrop is required');
+
+  const groupTitles = useMemo(() => data.map(({ groupTitle }) => groupTitle), [data]);
+
+  const [addGroupModalVisible, setAddGroupModalVisible] = useState(false);
+  const [isDisabled, setIsDisabled] = useState(false);
+
+  const onAddGroupRequest = async (e) => {
+    e.preventDefault();
+    const { groupTitle } = Object.fromEntries(new FormData(e.target));
+    if (!groupTitle) return toast.error('Le titre du groupe est obligatoire');
+    if (data.find((group) => group.groupTitle === groupTitle)) return toast.error('Ce groupe existe déjà');
+    await onAddGroup(groupTitle);
+    setAddGroupModalVisible(false);
+  };
+
+  const onDragAndDropRequest = useCallback(async () => {
+    const groupsElements = gridRef.current.querySelectorAll('[data-group]');
+    const groups = [...groupsElements].map((group) => group.dataset.group).map((groupTitle) => ({ groupTitle, items: [] }));
+    for (const group of groups) {
+      const categoriesElements = gridRef.current.querySelectorAll(`[data-group="${group.groupTitle}"] [data-item]`);
+      group.items = [...categoriesElements].map((category) => category.dataset.item);
+    }
+    if (groups.length !== data.length) {
+      capture('Drag and drop group error', { extra: { groups, data, title } });
+      return toast.error('Désolé, une erreur est survenue lors du glisser/déposer', "L'équipe technique a été prévenue. Vous pouvez réessayer");
+    }
+    if (
+      groups.reduce((allItems, group) => [...allItems, ...group.items], []).length !==
+      data.reduce((allItems, group) => [...allItems, ...group.items], []).length
+    ) {
+      capture('Drag and drop categories error', { extra: { groups, data, title } });
+      return toast.error('Désolé, une erreur est survenue lors du glisser/déposer', "L'équipe technique a été prévenue. Vous pouvez réessayer");
+    }
+    // /* there is a bug sometimes with the drag and drop, where some items are duplicated or even groups disappear...
+    //   we need to check that drag-n-drop only drag-n-dropped and didn't add/remove anything
+    // */
```

Comment:
Ah ouais ? Mince ! C'est la merde non ?

```suggestion
    // There is a bug sometimes with the drag and drop, where some items are duplicated or even groups disappear...
    //   we need to check that drag-n-drop only drag-n-dropped and didn't add/remove anything
```

Reply:
non mais en fait c'est réglé avec les lignes du dessus, on peut supprimer ce commentaire